### PR TITLE
stored: dird: add backup checkpoints that save backup metadata to the Catalog during the execution of the backup

### PR DESCRIPTION
**Pull request by alaaeddineelamri**
### Description

When doing backup jobs, backup metadata (number of files written to volume, number of bytes written, etc.) is only saved on a successful job. On a failed or canceled job, users end up with volumes filled with backed up data, but information about this data is not available in the catalog, and therefore already backed up data before the fail cannot be restored using the restore command. 
This PR addresses the issue by introducing checkpoints during a backup's execution that save necessary metadata (updating File, Job, and Jobmedia tables) that enables users to restore data even for failed or canceled jobs. Said checkpoints happen on volume switches, or after a certain interval of time set by the user in the SD configuration.

#### Please check

- [X] Short description and the purpose of this PR is present _above this paragraph_
- [X] Your name is present in the AUTHORS file (optional)

If you have any questions or problems, please give a comment in the PR.

### Helpful documentation and best practices

- [Git Workflow](https://docs.bareos.org/DeveloperGuide/gitworkflow.html)
- [Automatic Sourcecode Formatting](https://docs.bareos.org/DeveloperGuide/generaldevel.html#automatic-sourcecode-formatting)
- [Check your commit messages](https://docs.bareos.org/DeveloperGuide/gitworkflow.html#commits)

##### General
- [x] PR name is meaningful
- [x] Purpose of the PR is understood
- [x] Separate commit for this PR in the CHANGELOG.md, PR number referenced is same
- [x] Commit descriptions are understandable and well formatted
- [x] If backport: add original PR number and target branch at top of this file: **Backport of PR#000 to bareos-2x**

##### Source code quality

- [x] Source code changes are understandable
- [x] Variable and function names are meaningful
- [x] Code comments are correct (logically and spelling)
- [x] Required documentation changes are present and part of the PR
- [x] `bareos-check-sources --since-merge` does not report any problems
- [x] `git status` should not report modifications in the source tree after building and testing

##### Tests

- [x] Decision taken that a system- or unittest is required (if not, then remove this paragraph)
- [x] The decision towards a systemtest is reasonable compared to a unittest
- [x] Testname matches exactly what is being tested
- [x] Output of the test leads quickly to the origin of the fault

### CODE REVIEW

**Review by pstorz — CHANGES_REQUESTED** [review]
Very good work.

However, I see the following things that need to be tested in a system test:
1: Kill the filedaemon during the backup and check that the last successful checkpoint can be recovered.
2: The same with killing the storage daemon.
3: The same with the director.

4: We also need to measure the impact on backup speeed with a big amount data both to  disk and to tape.Depending on the results we can also define a suggested value for the checkpoint interval.

5: The documentation of the feature needs to be written based on the results of the testing..

File: core/src/stored/append.cc [hunk 1]
```diff
@@ -180,60 +319,80 @@ bool DoAppendData(JobControlRecord* jcr, BareosSocket* bs, const char* what)
...
-    break;
 
-  fi_checked:
     if (file_index != last_file_index) {
-      jcr->JobFiles = file_index;
       last_file_index = file_index;
+      if (file_currently_processed.GetFileIndex() > 0) {
+        processed_files.push_back(std::move(file_currently_processed));
+      }
+      file_currently_processed.Initialize(file_index);
```
Comment (arogge) [inline core/src/stored/append.cc:338]: the standard mandates that a moved-from object is "in a valid, but unspecified state" which means invoking `Initialize()` as a member-function raises a flag here.
It is probably nicer to have an explicit constructor that takes file_index and then do `file_currently_processed = FileProcessed{file_index}` which will move-assign.

File: core/src/stored/append.h [hunk 2]
```diff
@@ -21,11 +21,39 @@
...
+  ProcessedFile(const ProcessedFile& other);
+  ProcessedFile& operator=(const ProcessedFile& other);
+  ProcessedFile(ProcessedFile&& other);
+  ProcessedFile& operator=(ProcessedFile&& other);
+
+  void SendAttributesToDirector(JobControlRecord* jcr);
+  void Initialize(int32_t index);
+  void AddAttribute(DeviceRecord* record);
+  inline int32_t GetFileIndex() { return fileindex_; }
+  inline std::vector<DeviceRecord> GetAttributes() { return attributes_; }
```
Comment (arogge) [inline core/src/stored/append.h:45]: When defining the function inside the class' body, the `inline` is implied.
```suggestion
  int32_t GetFileIndex() { return fileindex_; }
  std::vector<DeviceRecord> GetAttributes() { return attributes_; }
```

Comment (arogge) [inline core/src/stored/append.h:45]: this will return a copy of the vector. Is that intended?

File: core/src/stored/append.h [hunk 3]
```diff
@@ -21,11 +21,39 @@
...
+
+
+ private:
+  int32_t fileindex_{-1};
+  std::vector<DeviceRecord> attributes_;
```
Comment (arogge) [inline core/src/stored/append.h:50]: sorry, but you simply cannot safely put a `DeviceRecord` in a `std::vector<>` right now, `DeviceRecord` is not prepared for that.

**Review by arogge — COMMENTED** [review]
I'm not happy with `ProcessedFile`, it is pretty error-prone right now and I'm not even sure it works correctly currently. However, the Problem  is mostly `DeviceRecord`. We either need to fix that or come up with a "better" object that we can construct from a `DeviceRecord` and that we can safely store in a vector.

File: core/src/stored/append.cc [hunk 4]
```diff
@@ -52,39 +50,165 @@ static char OK_replicate[] = "3000 OK replicate data\n";
...
+  this->attributes_ = std::move(other.attributes_);
+
+  other.fileindex_ = 0;
+
+  return *this;
+}
+
+void ProcessedFile::SendAttributesToDirector(JobControlRecord* jcr)
+{
+  for (auto attribute : attributes_) { SendAttrsToDir(jcr, &attribute); }
```
Comment (arogge) [inline core/src/stored/append.cc:90]:
```suggestion
  for (const auto& attribute : attributes_) { SendAttrsToDir(jcr, &attribute); }
```

**Review by arogge — COMMENTED** [review]
After giving it another thought, I'd probably do the following:
Add a new class (maybe `ProcessedFileData`) that only contains the members required for `SendAttrsToDir()` and `StorageDaemonDeviceControlRecord::DirUpdateFileAttributes()`. You can then use a temporary `DeviceRecord` in `ProcessedFile::SendAttributesToDirector()` that gets a copy of the required values (i.e. also a copy of the pointer to data) and is passed to `SendAttrsToDir()`.

Now if that new class has its data-member implemented with something like `std::vector<char>` or `std::string`, then destruction, copy and move will just work out of the box. This will simplify `ProcessedFile` as you don't need any special member function (rule-of-zero).
When the new class provides a constructor that takes a `const DeviceRecord&`, you can just implement `ProcessedFile::AddAttribute(const DeviceRecord& record)` as `attributes_.emplace_back(record);`

File: core/src/cats/sql_update.cc [hunk 5]
```diff
@@ -161,18 +174,33 @@ bool BareosDb::UpdateJobEndRecord(JobControlRecord* jcr, JobDbRecord* jr)
...
-      "ClientId=%u,JobBytes=%s,ReadBytes=%s,JobFiles=%u,JobErrors=%u,"
-      "VolSessionId=%u,"
-      "VolSessionTime=%u,PoolId=%u,FileSetId=%u,JobTDate=%s,"
-      "RealEndTime='%s',PriorJobId=%s,HasBase=%u,PurgedFiles=%u WHERE JobId=%s",
-      (char)(jr->JobStatus), (char)(jr->JobLevel), dt, jr->ClientId,
-      edit_uint64(jr->JobBytes, ed1), edit_uint64(jr->ReadBytes, ed4),
-      jr->JobFiles, jr->JobErrors, jr->VolSessionId, jr->VolSessionTime,
-      jr->PoolId, jr->FileSetId, edit_uint64(JobTDate, ed2), rdt, PriorJobId,
-      jr->HasBase, jr->PurgedFiles, edit_int64(jr->JobId, ed3));
+  if (jr->JobFiles > 0) {
```
[the remark this answers is not in the document]
Comment (alaaeddineelamri) [reply core/src/cats/sql_update.cc:181]: Yes, we need to handle this. When canceling a job, the director tries to receive the `JobFiles` from the FD, but as the connection drops because of the `cancel`, the FD is unable to send the data. So the container that was supposed to receive the JobFiles stays in its initialized state i.e. all zeros, and then the director proceeds to update the database with this empty data, which would cause it to overwrite the data sent from the SD during checkpoints.

And yes, the duplication is ugly, I've changed it.

File: core/src/cats/sql_update.cc [hunk 6]
```diff
@@ -100,16 +100,29 @@ bool BareosDb::UpdateJobStartRecord(JobControlRecord* jcr, JobDbRecord* jr)
...
        edit_int64(jr->ClientId, ed1), edit_uint64(JobTDate, ed2),
        edit_int64(jr->PoolId, ed3), edit_int64(jr->FileSetId, ed4),
-       edit_int64(jr->JobId, ed5));
+       jcr->VolSessionId, jcr->VolSessionTime, edit_int64(jr->JobId, ed5));
 
   changes = 0;
   return UPDATE_DB(jcr, cmd) > 0;
 }
 
+bool BareosDb::UpdateRunningJobRecord(JobControlRecord* jcr, JobDbRecord* jr)
```
[the remark this answers is not in the document]
Comment (alaaeddineelamri) [reply core/src/cats/sql_update.cc]: agreed.

File: core/src/stored/append.h (hunk 2 above)
Comment (alaaeddineelamri) [reply core/src/stored/append.h]: Yes, I use it for testing to check contents

File: core/src/stored/append.cc [hunk 7]
```diff
@@ -52,39 +50,151 @@ static char OK_replicate[] = "3000 OK replicate data\n";
...
+  UpdateFileList(jcr);
+  UpdateJobmediaRecord(jcr);
+  Jmsg(jcr, M_INFO, 0, _("Checkpoint completed\n"));
+}
+
+static time_t DoTimedCheckpoint(JobControlRecord* jcr,
+                                time_t checkpoint_time,
+                                time_t checkpoint_interval)
+{
+  time_t now
```
Comment (arogge) [inline core/src/stored/append.cc:134]:
```suggestion
  const time_t now
```

File: core/src/stored/append.cc [hunk 8]
```diff
@@ -52,39 +50,151 @@ static char OK_replicate[] = "3000 OK replicate data\n";
...
+  return checkpoint_time;
+}
+
+static void SaveFullyProcessedFiles(JobControlRecord* jcr,
+                                    std::vector<ProcessedFile>& processed_files)
+{
+  if (!processed_files.empty()) {
+    for_each(
+        processed_files.begin(), processed_files.end(),
+        [&jcr](ProcessedFile& file) { file.SendAttributesToDirector(jcr); });
```
Comment (arogge) [inline core/src/stored/append.cc:153]: when `SendAttributesToDirector()` is marked const, you can pass ProcessedFile as const&
```suggestion
    for_each(
        processed_files.begin(), processed_files.end(),
        [&jcr](const ProcessedFile& file) { file.SendAttributesToDirector(jcr); });
```

File: core/src/stored/append.cc [hunk 9]
```diff
@@ -52,39 +50,151 @@ static char OK_replicate[] = "3000 OK replicate data\n";
...
+       _("Checkpoint: Syncing current backup status to catalog\n"));
+  UpdateJobrecord(jcr);
+  UpdateFileList(jcr);
+  UpdateJobmediaRecord(jcr);
+  Jmsg(jcr, M_INFO, 0, _("Checkpoint completed\n"));
+}
+
+static time_t DoTimedCheckpoint(JobControlRecord* jcr,
+                                time_t checkpoint_time,
+                                time_t checkpoint_interval)
```
Comment (arogge) [inline core/src/stored/append.cc:132]: I would have expected to simply pass the next checkpoint time by reference, but your solution is good, too.
```
static time_t DoTimedCheckpoint(JobControlRecord* jcr,
                                time_t checkpoint_time,
                                time_t checkpoint_interval)
```

File: core/src/stored/append.cc [hunk 10]
```diff
@@ -52,39 +50,151 @@ static char OK_replicate[] = "3000 OK replicate data\n";
...
+{
+  time_t now
+      = std::chrono::system_clock::to_time_t(std::chrono::system_clock::now());
+  if (now > checkpoint_time) {
+    checkpoint_time = now + checkpoint_interval;
```
Comment (arogge) [inline core/src/stored/append.cc:137]: Is that what the user would expect?
If the checkpoint was delayed for some reason should the next checkpoint happen at
`now + checkpoint_interval`
or should it be more like
 `while(checkpoint_time <= now) checkpoint_time += checkpoint_interval;`

**Review by arogge — COMMENTED** [review]
I still have some suggestions for the C++ code, but as far as I can tell there is no major issue left (at least of those that I can find).
Hopefully I'll get to compile and test this later today.

File: core/src/stored/append.cc (hunk 10 above)
Comment (alaaeddineelamri) [reply core/src/stored/append.cc]: I see where you're going, I agree, your solution keeps the pace of checkpoints more regular looking

File: core/src/stored/append.cc (hunk 9 above)
Comment (alaaeddineelamri) [reply core/src/stored/append.cc:132]: I thought about it too, but I wanted the checkpoint time change to be kind of explicit

File: core/CTestScript.cmake.in [hunk 11]
```diff
@@ -20,8 +20,10 @@ set(CTEST_SOURCE_DIRECTORY @CMAKE_SOURCE_DIR@)
...
-ctest_test(EXCLUDE_LABEL broken
-           PARALLEL_LEVEL 5
-           @REPEAT_UNTIL@
-           )
+ctest_test(EXCLUDE_LABEL broken|flaky PARALLEL_LEVEL 5)
+ctest_test(
+  INCLUDE_LABEL flaky
+  PARALLEL_LEVEL 5
+  @REPEAT_UNTIL@
+)
```
Comment (arogge) [inline core/CTestScript.cmake.in:28]: Splitting this into two sets breaks Jenkins' ability to see the tests. It will only take the later run into account, which leads to a major regression is test evaluation.

File: core/src/tests/append_test.cc [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+class AppendProcessedFileTest : public testing::Test {
+ public:
+  int32_t arbitrary_index{1};
+  storagedaemon::ProcessedFile file{};
+  void SetUp() override
+  {
+    file = storagedaemon::ProcessedFile{arbitrary_index};
+  }
+};
+
```
[the remark this answers is not in the document]
Comment (arogge) [reply core/src/tests/append_test.cc:40]: Sorry, but I don't see how this was addressed. Can you point me in the right direction?

File: systemtests/tests/virtualfull-bscan/testrunner [hunk 13]
```diff
@@ -42,7 +42,7 @@ run job=$JobName level=Full yes
 wait
 messages
 list jobs
-@exec "sh -c 'touch ${tmp}/data/*.c'"
+@exec "sh -c 'touch ${tmp}/data/build/src/tests/*.c'"
```
Comment (arogge) [inline systemtests/tests/virtualfull-bscan/testrunner:45]: but that test previously ran flawlessly without this change - and now it is broken on my machine.
I don't understand why this was at all required.

File: systemtests/tests/virtualfull/testrunner [hunk 14]
```diff
@@ -43,7 +43,7 @@ run job=$JobName level=Full yes
 wait
 messages
 list jobs
-@exec "sh -c 'touch ${tmp}/data/*.c'"
+@exec "sh -c 'touch ${tmp}/data/build/src/tests/*.c'"
```
Comment (arogge) [inline systemtests/tests/virtualfull/testrunner:46]: see above.

**Review by arogge — DISMISSED** [review]

File: core/src/stored/dir_cmd.cc [hunk 15]
```diff
@@ -1029,7 +1030,8 @@ static bool MountCmd(JobControlRecord* jcr)
           } else {
             dir->fsend(
                 _("3905 Device %s open but no Bareos volume is mounted.\n"
-                  "If this is not a blank tape, try unmounting and remounting "
+                  "If this is not a blank tape, try unmounting and "
+                  "remounting "
```
Comment (pstorz) [inline core/src/stored/dir_cmd.cc:1034]: This line could be melted into the line before.

File: core/src/stored/stored_conf.cc [hunk 16]
```diff
@@ -92,6 +92,7 @@ static ResourceItem store_items[] = {
   {"SdConnectTimeout", CFG_TYPE_TIME, ITEM(res_store, SDConnectTimeout), 0, CFG_ITEM_DEFAULT, "1800" /* 30 minutes */, NULL, NULL},
   {"FdConnectTimeout", CFG_TYPE_TIME, ITEM(res_store, FDConnectTimeout), 0, CFG_ITEM_DEFAULT, "1800" /* 30 minutes */, NULL, NULL},
   {"HeartbeatInterval", CFG_TYPE_TIME, ITEM(res_store, heartbeat_interval), 0, CFG_ITEM_DEFAULT, "0", NULL, NULL},
+  {"CheckpointInterval", CFG_TYPE_TIME, ITEM(res_store, checkpoint_interval), 0, CFG_ITEM_DEFAULT, "0", NULL, NULL},
```
Comment (pstorz) [inline core/src/stored/stored_conf.cc:95]: Didn't we say that we wanted to enable the checkpoints by default?
I think the result of the tests was that even in big environments we could set it to 30 or 60 seconds, is that true?

File: docs/manuals/source/manually_added_config_directive_descriptions/sd-storage-CheckpointInterval.rst.inc [hunk 17]
```diff
@@ -0,0 +1 @@
+:index:`\ <single: Broken pipe>`\  This is the interval at which Backup checkpoints are executed. While the storage daemon is appending data to volumes, it does checkpoints that update backup Metadata in the Catalog.
```
Comment (pstorz) [inline docs/manuals/source/manually_added_config_directive_descriptions/sd-storage-CheckpointInterval.rst.inc:1]:
```suggestion
:index:`\ <single: checkpoint interval>`\  This is the interval at which Backup checkpoints are executed. When a checkpoint is executed, the files that have been successfully stored into the media at that point in time will be updated in the catalog database.
```

File: systemtests/tests/checkpoints/etc/bareos/bareos-dir.d/console/bareos-mon.conf.in [hunk 18]
```diff
@@ -0,0 +1,7 @@
+Console {
```
Comment (pstorz) [inline systemtests/tests/checkpoints/etc/bareos/bareos-dir.d/console/bareos-mon.conf.in:1]: Is this file required?

File: systemtests/tests/checkpoints/etc/bareos/bareos-fd.d/director/bareos-mon.conf.in [hunk 19]
```diff
@@ -0,0 +1,6 @@
+Director {
```
Comment (pstorz) [inline systemtests/tests/checkpoints/etc/bareos/bareos-fd.d/director/bareos-mon.conf.in:1]: Is this file required?

File: systemtests/tests/checkpoints/etc/bareos/bareos-sd.d/director/bareos-mon.conf.in [hunk 20]
```diff
@@ -0,0 +1,6 @@
+Director {
```
Comment (pstorz) [inline systemtests/tests/checkpoints/etc/bareos/bareos-sd.d/director/bareos-mon.conf.in:1]: This file probably also is not required.

File: systemtests/tests/checkpoints/testrunner-checkpoints-on-cancel [hunk 21]
```diff
@@ -0,0 +1,98 @@
...
+@$out /dev/null
+messages
+@$out $backup_log
+run job=$slowjob level=Full yes
+quit
+END_OF_DATA
+
+run_bconsole
+
+sleep 14
```
Comment (pstorz) [inline systemtests/tests/checkpoints/testrunner-checkpoints-on-cancel:37]: Waiting a fixed time is usually a bad idea. Better use a while loop with timeout.

File: systemtests/tests/checkpoints/testrunner-checkpoints-on-kill [hunk 22]
```diff
@@ -0,0 +1,87 @@
...
+@$out /dev/null
+messages
+@$out $backup_log
+run job=$slowjob level=Full yes
+quit
+END_OF_DATA
+
+    run_bconsole
+
+    sleep 17
```
Comment (pstorz) [inline systemtests/tests/checkpoints/testrunner-checkpoints-on-kill:41]: Same here, fixed wait time is bad, as it can fail on very busy systems or waste time on quick systems

File: systemtests/tests/checkpoints/testrunner-checkpoints-on-stop [hunk 23]
```diff
@@ -0,0 +1,92 @@
...
+@$out /dev/null
+messages
+@$out $backup_log
+run job=$slowjob level=Full yes
+quit
+END_OF_DATA
+
+run_bconsole
+
+sleep 14
```
Comment (pstorz) [inline systemtests/tests/checkpoints/testrunner-checkpoints-on-stop:38]: Again fixed wait time

File: systemtests/tests/parallel-jobs/etc/bareos/bareos-dir.d/console/bareos-mon.conf.in [hunk 24]
```diff
@@ -0,0 +1,7 @@
+Console {
```
Comment (pstorz) [inline systemtests/tests/parallel-jobs/etc/bareos/bareos-dir.d/console/bareos-mon.conf.in:1]: This file probably also can go.

File: systemtests/tests/parallel-jobs/etc/bareos/bareos-fd.d/director/bareos-mon.conf.in [hunk 25]
```diff
@@ -0,0 +1,6 @@
+Director {
```
Comment (pstorz) [inline systemtests/tests/parallel-jobs/etc/bareos/bareos-fd.d/director/bareos-mon.conf.in:1]: Not required?

File: systemtests/tests/stresstest/cloneandcopykernelcode [hunk 26]
```diff
@@ -0,0 +1,36 @@
...
+
+set -e
+set -o pipefail
+
+# clone the linux kernel code from github and make 10 copies of it
+# it would amount to 75GB of data, and around 1M files
+
+mkdir kernelcopies
+cd kernelcopies
+git clone https://www.github.com/torvalds/linux
```
Comment (pstorz) [inline systemtests/tests/stresstest/cloneandcopykernelcode:30]: Is this test enabled by default?
Probably we want to execute it only if explicitly asked for.

File: systemtests/tests/stresstest/etc/bareos/bareos-sd.d/director/bareos-mon.conf.in [hunk 27]
```diff
@@ -0,0 +1,6 @@
+Director {
```
Comment (pstorz) [inline systemtests/tests/stresstest/etc/bareos/bareos-sd.d/director/bareos-mon.conf.in:1]: This file probably can me removed.

File: systemtests/tests/stresstest/testrunner-stresstest [hunk 28]
```diff
@@ -0,0 +1,68 @@
...
+
+for i in {1..10}; do
+    rm -f $backup_log
+    run_bconsole
+    elapsedtime=$(grep 'Elapsed time:       ' $backup_log | sed -n -e 's/^.*time:       //p')
+    echo "$i,${elapsedtime}" >> $csvlog
+    rm storage/*
+done
+
+#cat << END_OF_DATA> "$tmp/bconcmds"
```
Comment (pstorz) [inline systemtests/tests/stresstest/testrunner-stresstest:50]: Quite a lot of commented out files here. Can they be removed?

**Review by pstorz — CHANGES_REQUESTED** [review]
Sorry for taking so long for reviewing it. 
The "mon" config files probably can be removed everywhere in the code, we can thin about doing that in a separate PR.

**Review by pstorz — APPROVED** [review]

**Review by pstorz — CHANGES_REQUESTED** [review]
there is some issue with the checkpoints-on-kill systemtest

**Review by pstorz — APPROVED** [review]
Good Work!
